### DISCUSSION TITLE
How to log request body using tracing

### DISCUSSION THREAD

**Discussion by davidbnk** [hyperium/tonic#1133] [Q&A] [answered]
I'm looking for a way to log the request body using tracing, so far I have arrived at this attempt:

```
Server::builder()
      .layer(
          tower_http::trace::TraceLayer::new_for_grpc()
              .make_span_with(DefaultMakeSpan::new().include_headers(true))
              .on_request(|request: &Request<Body>, _span: &Span| {
                  info!(
                      "request {} {} {}",
                      request.method(),
                      request.uri().path(),
                      request.body().data()
                  )
              }),
      )
      .add_service(my_service)
      .serve(addr)
      .await?;
```
But it doesn't work because `.body().data()` needs a mutable reference, and at this point in time the request just arrived and the body might not even be available

My next idea was maybe if I could have access to the Span in my service method where the request is available? I don't think its possible to have access to that tho. Is there a recommended approach of doing this?

**Comment by davidpdrsn** [comment, marked as answer]
This is hard because the body is an async stream that can only be consumed once.

I would recommend logging things in your handlers as necessary without relying on middleware.

You can probably also write a middleware that does it but it's not simple. I don't believe we have any examples of doing that.
